### PR TITLE
Add support for environment variables in source and class text fields in the plugin's configuration

### PR DESCRIPTION
**Pull request by NonGrate**
I've met the problem, that when environment variables are used in the JaCoCo configuration in Jenkins.

Exclusions variables is parsed and the value is sent to the JaCoCo report generation, but source and classes paths are passed to the report generation as strings.

This pull request fixes the bug and allows to use environment variables in the class paths and source paths test fields.

Here's the opened issue:
https://issues.jenkins-ci.org/browse/JENKINS-36168

### CODE REVIEW

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java [hunk 1]
```diff
@@ -447,13 +449,19 @@ public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath filePath, @Nonnull
         logger.print("[JaCoCo plugin] Saving matched execfiles: ");
         dir.addExecFiles(matchedExecFiles);
         logger.print(" " + Util.join(matchedExecFiles," "));
-        FilePath[] matchedClassDirs = resolveDirPaths(filePath, taskListener, classPattern);
+        FilePath[] matchedClassDirs = {};
+        if (run instanceof AbstractBuild) {
+            matchedClassDirs = resolveDirPaths((AbstractBuild) run, filePath, taskListener, classPattern);
+        }
```
Comment (centic9) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: Not sure about the Jenkins internas here, but shouldn't we keep the previous behaviour if run is not an AbstractBuild?

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java [hunk 2]
```diff
@@ -447,13 +449,19 @@ public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath filePath, @Nonnull
...
         logger.print("\n[JaCoCo plugin] Saving matched class directories for class-pattern: " + classPattern + ": ");
         for (FilePath file : matchedClassDirs) {
             dir.saveClassesFrom(file);
             logger.print(" " + file);
         }
-        FilePath[] matchedSrcDirs = resolveDirPaths(filePath, taskListener, sourcePattern);
+        FilePath[] matchedSrcDirs = {};
+        if (run instanceof AbstractBuild) {
+            matchedSrcDirs = resolveDirPaths((AbstractBuild) run, filePath, taskListener, sourcePattern);
+        }
```
Comment (centic9) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: dito

**Review by centic9 — CHANGES_REQUESTED** [review]
Please seem my comment/question. 

Also a basic unit-test would be good to verify and cover the new feature. You know, high coverage is good, that is what this plugin is all about :)

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java (hunk 1 above)
Comment (NonGrate) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: Done.

File: src/main/java/hudson/plugins/jacoco/JacocoPublisher.java (hunk 2 above)
Comment (NonGrate) [inline src/main/java/hudson/plugins/jacoco/JacocoPublisher.java]: Done.

**Comment by NonGrate** [comment]
Haven't wrote test yet. Planning to do that.

**Comment by don-vip** [comment]
@NonGrate are you still planning to finish the PR? I'd be glad to help.

**Pull request closed without merging**